test_partner_client.py

```diff
@@ -0,0 +1,131 @@
+
+        super().__init__()
+
+        if not isinstance(str, referral_code) and not len(referral_code) == 10:
+            raise AssertionError(
+                "referral_code needs to be 10 digit number string.")
+
+        if not (int(referral_code[:-1]) % 10 == int(referral_code[-1])):
+            raise AssertionError("Check code digit {} (last digit) did not match mod10 requierment" +
+                                 "in referral code: {}".format(referral_code[-1], referral_code))
+
+        self.referral_code = referral_code
+
+        if len(lab_referral) > 50:
+            raise AssertionError(
+                "Cannot have more then 50 characters in lab_referral code.")
+
+        # REVIEW When reviewing verify against docs that
+        #        this regex is actually correct.
+        allowed_chars_regex = re.compile(r"[^A-Za-z0-9\s\.-]")
+        if not bool(re.search(lab_referral)):
+            raise AssertionError(
+                "lab_referral can only contain A-Z, a-z, 0-9, white spaces, and -.")
+
+        self.lab_referral = lab_referral
+
+        def verify_date_format(date_str):
+            try:
+                datetime.strptime(date_str, '%Y%m%d')
+            except ValueError:
+                raise AssertionError(
+                    "Incorrect date format on {}, should be YYYYMMDD".format(date_str))
+
+        verify_date_format(arrival_date)
+        self.arrival_date = arrival_date
+
+        verify_date_format(result_date)
+        self.result_date = result_date
+
+        if len(comment) > 255:
+            raise AssertionError(
+                "comments cannot have more than 255 characters")
+
+        self.comment = comment
+
+        valid_covid_responses = ["negative", "positive", "failed"]
+
+        if cov19_result not in valid_covid_responses:
+            raise AssertionError("cov_19_result has to have one of the following values: {}".format(
+                ", ".join(valid_covid_responses)))
+
+    def get_as_dict(self):
+        return vars(self)
+
+
+class FailedInContactingTestPartner(Exception):
+    pass
+
+
+class TestPartnerClient(object):
+    # This is valid for v.6 for the parters API
+
+    def __init__(self, config):
+        # TODO Figure out how to configure this, given Claritys normal setup
+        super().__init__()
+        self._url = config.get("test_partner_url")
+        self._user = config.get("test_partner_user")
+        self._password = config.get("test_partner_password")
+
+    # TODO decide if we want to retry or not. It depends a little bit on the
+    #      rest of the scenario. This could potentially take us into a long chain
+    #      of retries. That would mean things take a long time before failing.
+    #      Don't know what is best here.
+    @retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+    def send_single_sample_result(self, test_partner_sample_info):
+
+        if not isinstance(TestPartnerSampleInformation):
```

Comment:
I don't know if we should be doing type checking in dynamic code, I doubt we'll do it consistently :) It will fail on the next line if it doesn't have `get_as_dict()` anyway.

Reply:
No. I just kind of wanted to make it clear that I wanted a validated instance here, since all the validation happens when building that class. Don't know it that was the best decision, though.